### PR TITLE
Robust Lists

### PR DESCRIPTION
**Pull request by CodaFi**
First off, SHAME!

Secondly, the eager implementation of this thing does not gel with the laziness of the data structure.  There's so much unnecessary evaluation of the list it makes it almost meaningless as a data structure compared to the built-in one.  We can "fix" that with recursion but then we'd lose both the magic of benign effects and any control over the performance of each of the combinators.

Rock and a hard place, eh?

### CODE REVIEW

**Comment by CodaFi** [comment]
I have a feeling we need to go fully recursive for a lot of these combinators (or use a lazier iterator).

**Comment by CodaFi** [comment]
:bulb: What if we cache length in the Cons case for O(1) access?  Infinite lists have a count of -1 internally and throw an exception when we try to take their length.  Infinite lengths propagate through appends or additional cons'es.

**Comment by CodaFi** [comment]
I'm quite happy with this right now.

Ready for review @pthariensflame

File: swiftz/List.swift [hunk 1]
```diff
@@ -6,72 +6,290 @@
...
+
 	/// Appends and element onto the front of a list.
-	static public func cons(h: A) -> List<A> -> List<A> {
-		return { t in List(h, t) }
+	public static func cons(head : A, tail : List<A>) -> List<A> {
+		return List(head, tail)
+	}
+
+	/// Destructures a list into its constituent parts or Nil.
+	public func match() -> ListMatcher<A> {
```
Comment (CodaFi) [inline swiftz/List.swift]: :warning: Breaking API Change :warning:

**Comment by pthariensflame** [comment]
Reviewing slowly via mobile.  Anyone else in @typelift/fp want to join in?

**Comment by pthariensflame** [comment]
Sorry for taking so long!  I've got no real problems with this.  :)

**Comment by CodaFi** [comment]
You never have to apologize for that kind of thing!  As long as I've got the greenlight, then :boat: so I can push more stuff through and hit 0.1.0 hard.

**Pull request merged**